### DISCUSSION TITLE
PS SPI bus stability questions with ESP32

### DISCUSSION THREAD

**Discussion by darthcloud** [johnbaumann/OpenMC#23] [Q&A] [answered]
Hi @johnbaumann ,

I think you probably know about my project BlueRetro.

On some system the SPI bus is unreliable and I often see received data from PS to be corrupted. See darthcloud/BlueRetro#102

I see you used the trick of stalling the first core0 on data rx/tx.
I also see you bit bang instead of using the ESP32 SPI peripheral.

Would you mind sharing the issues you observed and the solution you applied? :)

I could blindly implement it similar as you did and give it a try, but having a bit of context before I do so would be nice :).

Thanks in advance!
Jacques

**Comment by johnbaumann** [comment, marked as answer]
Hi @darthcloud ,

Will had reached out to me about this a few months back but we never really got a chance to discuss, I think he ended up talking in length with Sickle, whom is involved in xstation and has done some work with Thanos(mcpro) as well. Are you any closer to figuring out the source? I read through the open issue but it wasn't immediately clear if there was a smoking gun beyond certain console revisions being more affected.

I did indeed borrow your bare metal + core0 stall approach, thanks for that btw, hope you're cool with me borrowing it here :). Awesome job documenting things over on hackaday.io, big 'aha!' moment when I first read that article. I actually still need to apply your changes to this project to deal with newer ESP-IDF changes which broke the current older method.

To be honest, I just got fed up with trying to use the ESP's SPI driver(seems to be a common pain point in the ESP32 community). I couldn't get it to behave how I wanted, and I couldn't get the timing in line with what I was seeing with my logic analyzer, so bit-banging was really the only other option.

As for issues, If pads or cards take too long to respond, the console will either: Unselect the device and treat it as if it were removed, or more frequently, the communication will essentially run over it's time slot into the next device's communication. The latter will happen if you take too long to ACK. Could this be the random button press behavior you're seeing? Off the top of my head, ACK responses were usually 6-10uS after the last clock edge, and the line is asserted for 1-2uS. The BIOS(and games that use the bios card/pad handler) will sort of assert the select line low very briefly, un-assert, then re-select all over again to start the actual traffic. Later games seem to patch this out(I believe the patch was introduced sometime around PSYQ 4.3-4.5).

I handle communication with an interrupt on the negative edge, which then triggers a loop alternating between transceiving and processing data. Because of this happening during an interrupt I have checks at pretty much every point to see if select is still asserted. Additionally, since my device can be powered externally by USB, I have to check for long periods of inactivity so that we don't keep core0 stalled indefinitely when the console is powered off.

Sorry if I'm rambling or if anything is unclear, have had my head out of this project for a few months so having to dive back through to try and remember.

Best wishes,
John

**Reply by darthcloud** [reply]
Thanks a lot for taking time to answer :)

No problem in using stuff from BlueRetro.

I don't really got leads, my 3 PS1 and 2 PS2 work fine for me, and I suspect people having issues might not all be related to the same aspect. PS1 Digital might have some effect for some.

There probably 2 distinct issue a play simultaneously, I say this purely on past experience where I add weird issue that didn't make sense ;).

I'll try to stall core 0 on ps transaction and see what happen, I'll try bit-bang if that fail.
